frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/mobile_appflowy_date_picker.dart

```diff
@@ -18,146 +18,317 @@ import 'package:go_router/go_router.dart';
-              }
+            isRange: isRange,
+            selectedDay: dateTime,
+            startDay: isRange ? startDateTime : null,
+            endDay: isRange ? endDateTime : null,
+            focusedDay: focusedDateTime,
+            onDaySelected: (selectedDay) {
+              onDateSelectedFromDatePicker(selectedDay, null);
+            },
+            onRangeSelected: (start, end) {
+              onDateSelectedFromDatePicker(start, end);
+            },
+            onPageChanged: (focusedDay) {
+              setState(() => focusedDateTime = focusedDay);
             },
-            onRangeSelected: widget.onRangeSelected,
-            rebuildOnDaySelected: widget.rebuildOnDaySelected,
           ),
         ),
         const _Divider(),
-        if (widget.enableRanges && widget.onRangeChanged != null)
-          _EndDateSwitch(
+        if (widget.onIsRangeChanged != null)
+          _IsRangeSwitch(
             isRange: widget.isRange,
-            onRangeChanged: widget.onRangeChanged!,
+            onRangeChanged: (value) {
+              if (!isRange) {
+                justChangedIsRange = true;
+              }
+              widget.onIsRangeChanged!.call(value);
+              setState(() => isRange = value);
+            },
+          ),
+        if (widget.onIncludeTimeChanged != null)
+          _IncludeTimeSwitch(
+            showTopBorder: widget.onIsRangeChanged == null,
+            includeTime: includeTime,
+            onIncludeTimeChanged: (includeTime) {
+              widget.onIncludeTimeChanged?.call(includeTime);
+              setState(() => this.includeTime = includeTime);
+            },
           ),
-        _IncludeTimeSwitch(
-          showTopBorder: !widget.enableRanges || widget.onRangeChanged == null,
-          includeTime: _includeTime,
-          onIncludeTimeChanged: (includeTime) {
-            widget.onIncludeTimeChanged(includeTime);
-            setState(() => _includeTime = includeTime);
-          },
-        ),
         if (widget.onReminderSelected != null) ...[
           const _Divider(),
           _ReminderSelector(
-            selectedReminderOption: _reminderOption,
+            selectedReminderOption: reminderOption,
             onReminderSelected: (option) {
               widget.onReminderSelected!.call(option);
-              setState(() => _reminderOption = option);
+              setState(() => reminderOption = option);
             },
             timeFormat: widget.timeFormat,
             hasTime: widget.includeTime,
           ),
         ],
         if (widget.onClearDate != null) ...[
           const _Divider(),
-          _ClearDateButton(onClearDate: widget.onClearDate!),
+          _ClearDateButton(
+            onClearDate: () {
+              widget.onClearDate!.call();
+              Navigator.of(context).pop();
+            },
+          ),
         ],
         const _Divider(),
       ],
     );
   }
+
+  void onDateSelectedFromDatePicker(
```

Comment:
These two date pickers share similar logic. Should these functions, like `onDateSelectedFromDatePicker`, be composed together or managed using a bloc?

Reply:
For cubit/bloc, moving the logic to the cubit/bloc meant that triggering the callbacks required emitting extra states to have a BlocListener help do this. This resulted in a lot of boiler plate code in the widget and the cubit. And when the widget updates, that usually means the bloc is reinstantiated, or we need to update it separately

For an abstracted widget, it works a bit better, killing off ~120 lines. I think having an abstraction is overkill for this though and it makes the build methods hard to follow. What do you think of this? Or do you have something else in mind? https://github.com/AppFlowy-IO/AppFlowy/commit/9f682ca0d1a0b4e9af9662cc559f6d887597f289